src/v3/plugins/github/graphView.test.js

```diff
@@ -0,0 +1,383 @@
+          const authorsEdge = GE.createEdge.authors(userlike, issue);
+          (authorsEdge: any).dst = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(authorsEdge);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected dst");
+        });
+      });
+      describe("merged as edges", () => {
+        const commitAddress = NodeAddress.fromParts([
+          "sourcecred",
+          "git",
+          "COMMIT",
+          "hash",
+        ]);
+        it("src must be a pull", () => {
+          // $ExpectFlowError
+          const badEdge = GE.createEdge.mergedAs(issue, commitAddress);
+          failsForEdge(badEdge);
+        });
+        it("dst must be commit address", () => {
+          const badEdge = GE.createEdge.mergedAs(pull, GN.toRaw(repo));
+          failsForEdge(badEdge);
+        });
+        it("src must be pull in edge address", () => {
+          const otherPull = {type: "PULL", repo, number: "143"};
+          const mergedAs = GE.createEdge.mergedAs(otherPull, commitAddress);
+          (mergedAs: any).src = GN.toRaw(pull);
+          const g = exampleWithParents()
+            .addNode(commitAddress)
+            .addEdge(mergedAs);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected src");
+        });
+      });
+      describe("references edges", () => {
+        it("src must be a TextContentAddress", () => {
+          // $ExpectFlowError
+          const badEdge = GE.createEdge.references(userlike, pull);
+          failsForEdge(badEdge);
+        });
+        it("src must be the referrer in edge address", () => {
+          const references = GE.createEdge.references(issue, review);
+          (references: any).src = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(references);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected src");
+        });
+        it("dst must be referent in edge address", () => {
+          const references = GE.createEdge.references(issue, review);
+          (references: any).dst = GN.toRaw(pull);
+          const g = exampleWithParents().addEdge(references);
+          expect(() => new GraphView(g)).toThrow("Invariant: Expected dst");
+        });
+      });
+      describe("has parent edges", () => {
+        it("must satisfy specific hom relationships", () => {
+          const g = new Graph()
+            .addNode(GN.toRaw(repo))
+            // $ExpectFlowError
+            .addEdge(GE.createEdge.hasParent(repo, repo));
+          expect(() => new GraphView(g)).toThrow("Invariant: Edge");
+        });
+        it("must be unique", () => {
+          const otherRepo = {type: "REPO", owner: "foo", name: "bar"};
+          const g = exampleWithParents();
+          g.addNode(GN.toRaw(otherRepo));
+          const otherParent = {
+            src: GN.toRaw(issue),
+            dst: GN.toRaw(otherRepo),
+            address: EdgeAddress.append(GE._Prefix.hasParent, "foobar"),
```

Comment:
In fact, now that I think about this a little more: it seems to me that
uniqueness of parent edges is derivable from the invariants (a) every
child has at least one outgoing parent edge and (b) all parent edges
match the parent and child specified in the edge address. Maybe this is
what you were saying originally on Discord, and I didn’t understand. If
this is correct, then I’m fine removing the test case altogether if you
like.

Reply:
+1, except (b) all parent edges match the child address specified in the edge address